src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -242,10 +268,46 @@ public void Tick(DateTimeOffset now)
 
                 if (_timeoutAction == TimeoutAction.SendTimeoutResponse)
                 {
-                    Timeout();
+                    SetTimeoutResponse();
                 }
 
-                _frame.Stop();
+                Timeout();
+            }
+            else
+            {
+                lock (_readTimingLock)
+                {
+                    if (_readTimingEnabled)
+                    {
+                        _readTimingElapsed += timestamp - _lastTimestamp;
+
+                        if (_readTimingElapsed > _frame.RequestBodyTimeout.Ticks)
+                        {
+                            Log.RequestBodyTimeout(_context.ConnectionId, _frame.TraceIdentifier, TimeSpan.FromTicks(_frame.RequestBodyTimeout.Ticks));
+                            Timeout();
+                        }
+                        else if (_frame.RequestBodyMinimumDataRate?.Rate > 0)
+                        {
+                            var elapsedSeconds = (double)_readTimingElapsed / TimeSpan.TicksPerSecond;
+                            var rate = Interlocked.Read(ref _readTimingBytesRead) / elapsedSeconds;
+
+                            if (_readTimingElapsed > _frame.RequestBodyMinimumDataRate.GracePeriod.Ticks && rate < _frame.RequestBodyMinimumDataRate.Rate)
+                            {
+                                Log.RequestBodyMininumDataRateNotSatisfied(_context.ConnectionId, _frame.TraceIdentifier, _frame.RequestBodyMinimumDataRate.Rate);
+                                Timeout();
+                            }
+                        }
+
+                        // PauseTimingReads() cannot just set _timingReads to false. It needs to go through at least one tick
+                        // before pausing, otherwise _readTimingElapsed might never be updated if PauseTimingReads() is always
+                        // called before the next tick.
+                        if (_readTimingPauseRequested)
+                        {
+                            _readTimingEnabled = false;
```

Comment:
What happens if both PauseTimingReads and ResumeTimingReads are called between a single tick?.